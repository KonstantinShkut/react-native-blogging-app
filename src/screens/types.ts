import { PostItem } from '../services/blog/types';
import { StackScreenProps } from '@react-navigation/stack';

export const HOME = 'Home';
export const LOGIN = 'Login';
export const POST = 'Post';
export const POST_FORM = 'PostForm';
export const PROFILE = 'Profile';
export const SIGN_UP = 'SignUp';

type RootStackParamList = {
  [HOME]: undefined;
  [LOGIN]: undefined;
  [PROFILE]: undefined;
  [SIGN_UP]: { isTryingToPost: boolean } | undefined;
  [POST]: {
    id?: string;
    post?: PostItem;
  };
  [POST_FORM]: {
    post?: PostItem;
  };
};

export type ScreenProps<T extends keyof RootStackParamList> = StackScreenProps<
  RootStackParamList,
  T
>;
