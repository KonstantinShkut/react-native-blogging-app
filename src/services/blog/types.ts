import { FirebaseFirestoreTypes } from '@react-native-firebase/firestore';

export interface PostData {
  author: string;
  author_id: string;
  body: string;
  cover: string;
  published: FirebaseFirestoreTypes.Timestamp;
  title: string;
}

export interface PostItem extends PostData {
  id: string;
}
