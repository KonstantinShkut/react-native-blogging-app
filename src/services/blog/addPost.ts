import firestore, {
  FirebaseFirestoreTypes,
} from '@react-native-firebase/firestore';
import { PostData } from './types';

export const addPost = async (
  postData: PostData,
): Promise<FirebaseFirestoreTypes.DocumentReference> => {
  return await firestore().collection('posts').add(postData);
};
