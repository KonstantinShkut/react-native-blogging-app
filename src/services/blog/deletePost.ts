import firestore from '@react-native-firebase/firestore';
import { PostItem } from './types';

export const deletePost = async (post: PostItem) => {
  return await firestore().collection('posts').doc(post.id).delete();
};
