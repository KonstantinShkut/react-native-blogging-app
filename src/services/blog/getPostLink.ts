import { PostItem } from './types';

export const getPostLink = (post: PostItem) =>
  `bloggingapp://post/${post.id}`;
