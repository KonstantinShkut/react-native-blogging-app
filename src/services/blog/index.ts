export * from './addPost';
export * from './getPostLink';
export * from './loadPost';
export * from './loadPosts';
export * from './subscribeToPosts';
export * from './transformPostDoc';
