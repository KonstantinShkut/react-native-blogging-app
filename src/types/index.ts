export * from './PostItem';
export * from './PostData';
export * from './User';
