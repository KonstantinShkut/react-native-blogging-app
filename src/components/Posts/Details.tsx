import React from 'react';
import { Image, ScrollView, StyleSheet, Text, View } from 'react-native';
import { useSafeAreaInsets } from 'react-native-safe-area-context';
import moment from 'moment';
import {
  colorSecondary,
  container,
  coverImage,
  growingContainer,
  textHeader,
} from '../../styles';
import { ButtonRegular } from '../ButtonRegular';
import { PostItem } from '../../services/blog/types';

interface Props {
  post: PostItem;
  onShare: () => void;
}

export const Details = ({ post, onShare }: Props) => {
  const insets = useSafeAreaInsets();
  const { title, body, author, cover, published } = post;

  return (
    <ScrollView
      keyboardShouldPersistTaps="handled"
      showsVerticalScrollIndicator={false}
      style={container}
      contentContainerStyle={{
        ...styles.content,
        paddingBottom: insets.bottom,
      }}>
      <View>
        <Text style={textHeader}>{title}</Text>
        <Text style={styles.textAuthor}>
          {author} on {moment(published && published.toDate()).format('LL')}
        </Text>
        {!!cover && <Image source={{ uri: cover }} style={coverImage} />}
        <Text style={styles.textBody}>{body}</Text>
      </View>
      <View style={styles.bottom}>
        <ButtonRegular color={colorSecondary} title="Share" onPress={onShare} />
      </View>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  contentContainer: {
    padding: 20,
  },
  content: {
    ...growingContainer,
    padding: 20,
  },
  bottom: {
    flex: 1,
    justifyContent: 'flex-end',
    paddingBottom: 20,
  },
  textAuthor: {
    marginTop: 5,
    marginBottom: 20,
  },
  textBody: {
    fontSize: 18,
    marginVertical: 20,
  },
});
