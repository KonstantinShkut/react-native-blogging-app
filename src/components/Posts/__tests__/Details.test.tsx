import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import safeAreaContext from 'react-native-safe-area-context';
import navigation from '@react-navigation/native';
import { Details } from '../Details';
import { PostItem } from '../../../services/blog/types';

const mockFirebaseTimestamp = { toDate: () => new Date('2020-01-01') };
jest.mock('@react-navigation/native', () => jest.fn());
// @ts-ignore
navigation.useNavigation = jest.fn(() => ({ navigate: jest.fn() }));
jest.mock('react-native-safe-area-context', () => jest.fn());
// @ts-ignore
safeAreaContext.useSafeAreaInsets = jest.fn(() => ({
  bottom: 0,
}));

describe('Details', () => {
  it('renders correctly', () => {
    const post = {
      id: 'string-id',
      author: 'author',
      author_id: 'author-id',
      body: 'Post Body',
      cover: 'cover-url',
      published: mockFirebaseTimestamp,
      title: 'Post Title',
    };
    const component = renderer.create(
      <Details post={post as PostItem} onShare={() => null} />,
    );
    expect(component).toMatchSnapshot();
  });
});
