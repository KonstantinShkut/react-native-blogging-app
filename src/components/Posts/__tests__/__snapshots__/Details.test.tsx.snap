// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details renders correctly 1`] = `
<RCTScrollView
  contentContainerStyle={
    Object {
      "flexDirection": "column",
      "flexGrow": 1,
      "justifyContent": "center",
      "padding": 20,
      "paddingBottom": 0,
    }
  }
  keyboardShouldPersistTaps="handled"
  showsVerticalScrollIndicator={false}
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
>
  <View>
    <View>
      <Text
        style={
          Object {
            "fontSize": 33,
            "fontWeight": "bold",
          }
        }
      >
        Post Title
      </Text>
      <Text
        style={
          Object {
            "marginBottom": 20,
            "marginTop": 5,
          }
        }
      >
        author
         on 
        December 31, 2019
      </Text>
      <Image
        source={
          Object {
            "uri": "cover-url",
          }
        }
        style={
          Object {
            "borderColor": "#DDDDDD",
            "borderRadius": 10,
            "borderWidth": 1,
            "height": 200,
          }
        }
      />
      <Text
        style={
          Object {
            "fontSize": 18,
            "marginVertical": 20,
          }
        }
      >
        Post Body
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "justifyContent": "flex-end",
          "paddingBottom": 20,
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 10,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#607D8B",
                "borderColor": "#2089dc",
                "borderRadius": 3,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontSize": 18,
                  "paddingBottom": 1,
                  "paddingTop": 2,
                  "textAlign": "center",
                }
              }
            >
              Share
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
